### DISCUSSION TITLE
Possible CAD engine alternatives

### DISCUSSION THREAD

**Discussion by je-cook** [Fusion-Power-Plant-Framework/bluemira#1961] [General]
We currently use FreeCAD as our CAD engine but we should keep up with changes in the open source CAD world because it seems to be moving quite fast recently. Most are in the alpha/beta stage but options to be aware of:

- CADQuery, https://github.com/CadQuery/cadquery  - OCCT based (as is FreeCAD) uses a 'fluent' API instead of a direct API like FreeCAD. 
  **Update**: New direct API available and backend available in Bluemira with the environment variable `BLUEMIRA_GEOMETRY_BACKEND="cadquery"`
- Solvespace, https://solvespace.com, https://github.com/solvespace/solvespace - A C++ based BREP kernel
- _archived June 2026_ ~Fornjot, https://github.com/hannobraun/Fornjot - Rust based BREP CAD program, their CAD kernel is a separate crate~ 
- Truck, https://github.com/ricosjp/truck - Rust based BREP CAD program
- CADSketcher, https://github.com/hlorus/CAD_Sketcher - A Blender addon to try and bring CAD to the Blender ecosystem (uses solvespace)
- Ayam https://ayam.sourceforge.net, still maintained and regularly updated, since 2001

**Comment by je-cook** [comment]
these also look quite interesting: 
https://github.com/dune3d/dune3d
https://github.com/fogleman/sdf

**Comment by janlyy** [comment]
I would also add build123d https://github.com/gumyr/build123d to the list of alternatives

**Reply by je-cook** [reply]
Hi @janlyy, thanks for the addition, it was one I missed. 

With the reasonably recent addition of a direct API to CADQuery do you have any notes on the differences between the two? 
As far as I'm aware build123d was originally a fork of CADQuery with a direct API. 

As a side note we recently added CADQuery as an alternative backend to FreeCAD to take advantage of their new direct API (and lower level OCCT access). This can be tested out by setting the `BLUEMIRA_GEOMETRY_BACKEND` environment variable:
```export BLUEMIRA_GEOMETRY_BACKEND="cadquery" ```
